### PR TITLE
Consumer Tag

### PR DESCRIPTION
**Pull request by rlavolee**
adds the option to define a custom consumer tag or fallback to random tag

### CODE REVIEW

File: modules/client/src/main/scala/MessagingAPI.scala
Comment (hnaderi) [inline modules/client/src/main/scala/MessagingAPI.scala:47]: I think it's better to have an optional ctag with `None` default, which is the default incremental tags, while allowing users to pass a ctag when they want to override the name.

File: modules/protocol/src/main/scala/Domains.scala
Comment (hnaderi) [inline modules/protocol/src/main/scala/Domains.scala:105]: Creating a UUID is a side-effect and shouldn't be in a pure function, it also requires external dependencies to make it cross platform. The cats-effect way of doing it in a cross-platform way without requiring external dependencies is by using `UUIDGen[F]`.

I think it's better to leave it to be a user side decision, as there are no restrictions regarding what's the value of the ctag.

**Review by hnaderi — CHANGES_REQUESTED** [review]

File: modules/protocol/src/main/scala/Domains.scala
Comment (rlavolee) [reply modules/protocol/src/main/scala/Domains.scala:105]: Ok I'll remove UUID and get back the `consumer-x` as ctag default.

File: modules/client/src/main/scala/MessagingAPI.scala
Comment (rlavolee) [reply modules/client/src/main/scala/MessagingAPI.scala:47]: Ok so `consumeRaw` should look like this 

```scala
  def consumeRaw(
      queue: QueueName,
      noLocal: NoLocal = false,
      noAck: NoAck = true,
      exclusive: Boolean = false,
      arguments: FieldTable = FieldTable.empty,
      consumerTag: Option[ConsumerTag] = None
  ): Stream[F, DeliveredMessageRaw]
```

Should  `ctag` in `final def consume` also be optional ?

Comment (hnaderi) [reply modules/client/src/main/scala/MessagingAPI.scala]: Yes, everywhere the `ctag` is added should be optional to maintain backward source compatibility.

Comment (rlavolee) [reply modules/client/src/main/scala/MessagingAPI.scala]: I made it optional in my last commit

**Comment by rlavolee** [comment]
Hi @hnaderi !

I'm trying to get the PR done but I'm struggling with Mima.
Mima is complaining that `consumerRaw` is not there anymore because I added `consumerTag: Option[ConsumerTag] = None` in the signature.

```
method consumeRaw(java.lang.String,Boolean,Boolean,Boolean,scala.collection.immutable.Map)fs2.Stream in class lepus.client.Channel#ConsumingImpl does not have a correspondent in current version
```

I thought to add  the old `consumerRaw` back as shown below and call the new one in its body.

```scala
  def consumeRaw(
      queue: QueueName,
      noLocal: NoLocal = false,
      noAck: NoAck = true,
      exclusive: Boolean = false,
      arguments: FieldTable = FieldTable.empty,
      consumerTag: Option[ConsumerTag] = None
  ): Stream[F, DeliveredMessageRaw]

  final def consumeRaw(
      queue: QueueName,
      noLocal: NoLocal = false,
      noAck: NoAck = true,
      exclusive: Boolean = false,
      arguments: FieldTable = FieldTable.empty
  ): Stream[F, DeliveredMessageRaw] =
    consumeRaw(queue, noLocal, noAck, exclusive, arguments, None)
``` 

However the compiler is complaining about overloaded methods because of default arguments.

```
two or more overloaded variants of method consumeRaw have default arguments
```

What's the best way to deal with this situation ? Should I remove default args, break binary comp or introduce a new method name ? 🤔

**Comment by hnaderi** [comment]
@rlavolee 

> What's the best way to deal with this situation ? Should I remove default args, break binary comp or introduce a new method name ? 🤔

One way would be to add new methods instead of changing the old ones, but it's not worth it and I think we can break the bin-compat here, as the changes are source compatible.

I just bumped up the minor version and pushed it on the PR, so it should be fixed.

**Comment by hnaderi** [comment]
@rlavolee By the way, are you finished by the PR? or are there other changes you want to make?
If everything's OK, we can merge it. I'm just curios about testing a snapshot version on some real projects to see the behavior before merging, which I'll do tomorrow.

**Comment by rlavolee** [comment]
@hnaderi I'm done with it :) and eager to test it.

**Comment by hnaderi** [comment]
@rlavolee The snapshot version should be published in a few minutes.
Thanks for the contribution! :pray:

**Comment by rlavolee** [comment]
@hnaderi Your're welcome that was fun. Lepus is great !

**Review by gnoireaux — COMMENTED** [review]
LGTM 😉

And thanks for Lepus !

**Pull request merged**
